### ISSUE TITLE
Form onSubmit continues to pass in undefined...

### ISSUE THREAD

**Issue by fairbairn** [davidkpiano/react-redux-form#155]
Pretty simple scenario, but in every case, we print out `undefined` on the console at time of submission.

The model data is actually updated in the state tree within Redux.

```
class SampleForm extends Component {
    render() {
        return (
            <Form model="test"
                  onSubmit={ (model) => console.log(model) }>
                
                <Field model="test.name">
                    <input type="text"/>
                </Field>
                
                <button type="submit">Submit</button>
            </Form>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        test: state.form.test
    };
};

export default connect(
    mapStateToProps
)(SampleForm);
```

Are we missing something?

We're using this version...

`    "react-redux-form": "^0.10.5",`

```

**Comment by davidkpiano** [comment]
So the model is actually located at `form.test` in the store, meaning your `<Form>` should look like this:

```js
<Form model="form.test"
  // ... etc.
>
```

That should work, let me know! 😄

**Comment by fairbairn** [comment]
Your combination did not work...it submits, but it won't update the name item in the model.  In fact, it has no awareness of it.

This combination does...

```
class SampleForm extends Component {
    render() {
        return (
            <Form model="form.test"
                  onSubmit={ (model) => console.log(model) }>

                <Field model="test.name">
                    <input type="text"/>
                </Field>

                <button type="submit">Submit</button>
            </Form>
        )
    }
}
```

I believe the Form model should be at the same "level" as Field.  Field only works from the mapStateToProps tier down, but Form looks like it has to reference the model from the top of the Redux state tree down.

Seems they should be the same, right?

**Comment by fairbairn** [comment]
Pretty sure the issue is that the selectors are mis-aligned between Form and Field..

Here they are...

**Form**
```
function selector(state, { model }) {
  const modelString = typeof model === 'function'
    ? model(state)
    : model;

  return {
    ...state,
    modelValue: _get(state, modelString),
    formValue: getForm(state, modelString),
  };
}
```
**Field**
```
function selector(state, { model }) {
  const modelString = typeof model === 'function'
    ? model(state)
    : model;

  return {
    model: modelString,
    modelValue: _get(state, modelString),
  };
}
```
Field actually redefines model based on the assumed selector used.
Form takes it literally.

**Comment by davidkpiano** [comment]
Weird, all of the tests have it all at the same level. Can you share what your store looks like?

**Comment by fairbairn** [comment]
our basic store looks like this..

```
export const initialState = Object.assign({},
    {
        router: {},
        form: {},
        ui: {
            toggle_state: false
        }
    }
);
```

However, we provide a helper function that puts custom forms under a top level form so we can nest with this..

```
import {combineReducers} from 'redux';
import {modelReducer, formReducer} from 'react-redux-form';

const form_sections = [
    {
        form: 'test',
        fields: [
            'sample_text',
            'sample_radio',
            'sample_checkbox'
        ]
    }
];

function buildDynamicReducers(sections) {
    return sections.reduce((combined, section, index) =>{
        let name = section.form;
        combined[name] = modelReducer(name, {});
        combined[`${name}_form`] = formReducer(name);
        return combined;
    }, {})
}

const reducers = buildDynamicReducers(form_sections);

export default combineReducers(reducers);
```

That shouldn't matter because the state tree is perfect. 

If we provide it top level like you stated, then it doesn't find or auto-create the model field value in the state tree.  If we do it our way, it does.

Unrelated it seems (because we've tried both ways), we can't get checkboxes to work.  It sets it true, but never actually checks the checkbox, and clicking again doesn't set it to false.

Pretty simple syntax for a checkbox...

```
                        <Field model={model}>
                            <label>
                                <input type="checkbox"/> {label}
                            </label>
                        </Field>
```

**Comment by fairbairn** [comment]
For example, this doesn't ever check the checkbox...yet it sets the state variable of form.test.toggle to true.  It doesn't go false though.

```
class SampleForm extends Component {
    render() {
        return (
            <form>
                <Field model="test.toggle">
                    <input type="checkbox"/> Something
                </Field>
            </form>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        test: state.form.test
    };
};

export default connect(
    mapStateToProps
)(SampleForm);
```

**Comment by davidkpiano** [comment]
@fairbairn Take a look at this gist: http://esnextb.in/?gist=37f259673130c6c6ed8076d73ca3afca

Press **Execute**, play around with the form, and open your console. Both the `<Form>` and the `<Field>` need to have `model="test"` and `model="test.name"`, respectively.

I'll look into the checkbox issue.

**Comment by fairbairn** [comment]
Here's a modified version of yours. I changed one sequence, and state still works, but the submit returns undefined.  I also too added a checkbox, and it does not work.

http://esnextb.in/?gist=d21c090976c1b80f393c668d9ea447a0

Specifically, what I did is **nest the reducers**...(which is appropriate for redux)...so in my scenario, I took the exact test and test_form nodes and nested them one deep under form.  Your system tracks the state of the value properly.  

```
const reducers = buildDynamicReducers(form_sections);

const moreReducers = {
  form: combineReducers(reducers)
}
const store = applyMiddleware(thunk, createLogger())(createStore)
  (combineReducers(moreReducers));
```

**Comment by fairbairn** [comment]
Another version, this one leveraging connect..

http://esnextb.in/?gist=d21c090976c1b80f393c668d9ea447a0

Same nesting.  Same results.

**Comment by davidkpiano** [comment]
Okay, here is the updated gist from your example:

http://esnextb.in/?gist=fefd5b45add579274a9ada5a0ae884b8

The issue is that react-redux-form needs to know _exactly_ where the model lives in the root store, for both the `<Form>` and the `<Field>`. You were specifying it as living in e.g. `model="test.name"` even though it really lived in `model="form.test.name"`.

Same issue with your `buildDynamicReducers` function:

```js
function buildDynamicReducers(sections) {
  return sections.reduce((combined, section, index) =>{
    let name = section.form;
-   combined[name] = modelReducer(name, {});
+   combined[name] = modelReducer(`form.${name}`, {});
    combined[`${name}_form`] = formReducer(name);
    return combined;
  }, {})
}
```

Hope that helps!

**Comment by fairbairn** [comment]
Makes sense.  Guess my next question is since we have to be explicit with the reducers, and since we have to reference model in both cases from the root of state, do we need to connect at all? or are we simply always going to connect(s => s)?

**Comment by davidkpiano** [comment]
> It's also a little confusing, because the modelReducer has to be referenced from the root of state, but the formReducer does not?

Right; every model is based on the `modelReducer`, and the `formReducer` can live anywhere.

> do we need to connect at all?

Nope! Unless you need to grab a value from the state in your component, each `<Field>, <Form>, etc.` is already `connect()`ed on its own. They are each isolated and have no knowledge of their parent component.

**Comment by fairbairn** [comment]
I would argue that the inputs (in terms of a model reference) to the modelReducer and the formReducer should be the same.  People will pull their hair out otherwise.  The fact that these reducers are created in a nested reducer is what tags it to state.form in the first place.  I have no problem that the modelReducer and formReducer need to be told where to look.  It just seems confusing to tell them differently.

I can hack around this though and I greatly appreciate your help.

I do have a question why validators and errors both exist, vs. one single validation grouping that handles the state (true/false) and any error messages that are releated.  Did .errors come later and you didn't want to break backwards compatibility?

Just curious.

Once Form supports the errors prop and submit characteristic, I think we we're good to go.

Thanks!!!

**Comment by fairbairn** [comment]
Back in business!!  Awesome.

**Comment by davidkpiano** [comment]
Thanks! Feel free to propose any enhancements you'd like in separate tickets; I'd love to discuss them further in depth with you.

**Comment by fairbairn** [comment]
I will!

I actually have a slightly different philosophy on how errors could be handled.  I'm working around your structure, but in almost every forms engine I've ever worked with, the errors (per field) usually come back as a list of messages.  In your system, each message essentially has to have a key, although there's no real value in the key (unless I guess you needed to pinpoint specific messages or error conditions), so in our case we have to manufacturer one.  

No biggie.  We're using a central validation system, so there is no keyed validation tests.  Instead, we validate the field across a number of tests with a single function that's coded.  So for right now, we're just making the `errors: { reasons: fun }` keyed event be called `reasons`.

We can't use `validators` because we need those error messages from our validation engine, hence the need for `errors` support at the `<Form>` component level so we can just validate on submit and then have the messages available to us for display purposes.

Getting closer though to our ideal, and greatly appreciate your rapid response.

**Comment by davidkpiano** [comment]
> In your system, each message essentially has to have a key

Not necessarily; you can `setErrors(model, ['error 1', 'error 2'])` to an array. RRF is agnostic to this. Have you tried this?

And the reason for `validators` and `errors` both coexisting is because they are necessary for different use cases. For example, I can express my validators as this:

```js
<Field model="user.email"
  validators={{ isEmail, isRequired, (val) => val.length > 6 }}>
```

which makes it very concise. I can also set error messages from the API, which returns them as a dict of messages as well.

Another good reason for validators is **internationalization** - I do _not_ want to have i18n logic in my validators, nor do I want hardcoded messages, and I know many developers have the same concerns. A couple of the apps I'm building are working with i18n messages.

**Comment by fairbairn** [comment]
So a guidance question...

If we're going to build our own errors handler from the top-level Form, intercepting our submit handler, and we want to iterate through all the fields (I hate using the word model, model to me sounds like all of the fields, not a single field, so bear with me), and we want to file away an errors array of messages for each field if our validation function detects some failures (or it may not), should we be looking at `setErrors` or `setFieldsErrors`?  Keeping in mind that at the field rendering layer, we're going to need to access this again by field name so we can handle the presentation properly, by field.  We actually do both.  We show field level indicators as well as top level aggregate presentation.

I get a bit confused why there's an errors block under the formReducer section, but there's also an errors block under the fields individually as well - guess the top one are form errors and the fields are field specific?  

Based on that assumption, my guess would be we care about `setFieldsErrors` instead.  Finally, our validation routine may not have any tripped errors, in which case our array of error messages would be empty `[]`.  Do we need to call a `resetErrors` before we run through our routine to essentially set it back to a valid status before we check?  and would that touch the errors under fields or just the form level errors?

We can dispatch any number of events when our validation routine runs - just need to know which of the broad array of functions we should really care about depending on desired granularity.

**Comment by fairbairn** [comment]
Ugh, I think we have the same issue with `actions.setErrors` and `actions.setFieldsErrors`, a referencing issue relating to the model name when you get nested...

Again, our state has the model at form.test, and the form at form.test_form, so our model and form reducers are nested under state.form with a form name of `test`.

So this time, I'm firing this..

```
            dispatch(actions.setFieldsErrors(name, errors));
```

Where name is `loan_amount` and errors is an array of error messages properly calculated from validating the field.

The state has this model at `state.form.test.loan_amount`, but the only way I could get the action to fire correctly was to have `name` as just `loan_amount`.  I tried 'form.test.loan_amount' and 'test.loan_amount' and it didn't set a validity result.  Just the field name seemed to work, kind of..

That resulted in this happening in Redux...

![screenshot_2016-04-21_20-54-55](https://cloud.githubusercontent.com/assets/5289625/14728503/643c6ed0-0803-11e6-87e3-893e917b06d7.png)

Unfortunately, nothing gets affected under state.form.test_form (validity, errors, or fields).  I was expecting state.form.test_form.fields to be populated, but it was not.

**Comment by fairbairn** [comment]
I think I figured out part of the issue...

Interestingly, per my original comment, the formReducer DOES have to match with the modelReducer, so in my case, both values must be `form.test`.

**Comment by davidkpiano** [comment]
Hmm, this is a lot to digest, so let's continue the route of demonstrating the use case using esnextb.in , if you can... But I'll answer some of the questions you posed:

> `form: { ... model: "test" }`

First thing I noticed is this, which should probably be `model: "form.test"`. Think of `model` as saying "where in my state should I look for this value?" This alone will solve most (if not all) of your issues.

I am strongly considering implementing middleware to ease this burden - let me know if you think this is a good idea.

> If we're going to build our own errors handler from the top-level Form, intercepting our submit handler, and we want to iterate through all the fields, and we want to file away an errors array of messages for each field if our validation function detects some failures (or it may not), should we be looking at setErrors or setFieldsErrors?

You should be looking at `setFieldsErrors` to set all errors at once, which would look like this:

```js
dispatch(setFieldsErrors('form.test', {
  name: ['contains illegal chars', 'is too short'],
  email: ['is invalid email']
});
```

> I get a bit confused why there's an errors block under the formReducer section, but there's also an errors block under the fields individually as well - guess the top one are form errors and the fields are field specific?

Precisely. Just as you would construct your state tree in Redux such that if a state references two leaf-node reducers, it should probably be in a reducer higher up that's common to both reducers, you'd do the same for errors.

For example, you might have a `"passwords_match"` validation key that depends on two fields - `"password"` and `"confirm_password"`. This would likely be a form-level validation key.

> Finally, our validation routine may not have any tripped errors, in which case our array of error messages would be empty []. Do we need to call a resetErrors before we run through our routine to essentially set it back to a valid status before we check? and would that touch the errors under fields or just the form level errors?

Not after I push the patch that fixes #151. Then dispatching `setErrors(model, [])` would replace the errors of that `model` with an empty array. which would signify a valid field.

> We can dispatch any number of events when our validation routine runs - just need to know which of the broad array of functions we should really care about depending on desired granularity.

Since you're caring about error messages, you essentially only need to know about one - `setErrors(model, errors)`. The `setFieldsErrors(model, fieldsErrorsMap)` is a convenience action creator that batch executes `setErrors` for each key in the `fieldsErrorsMap` in the reducer. It's a good idea to use `setFieldsErrors` for performance, as it only triggers a single rerender as opposed to _n_ rerenders for each field.

**Comment by fairbairn** [comment]
Thank you David.  This is helpful, although I have to say, I went into this extremely enthusiastic about your library, but now I'm just frustrated.

I think there's a lot here under the hood, and maybe the issue is simply competing docs and yet-to-be-published docs while pre-1.0 is still in so much flux - there's certainly a *right way* to tackle a particular issue, but it's difficult to intuit that if you begin to really segment out logic and dig into how to leverage it in a modular way.  I really didn't think us moving out our model and model_form nodes from the root of state would cause so many challenges.  We're going to have hundreds of forms in our app.  I really didn't want them all at the root of state.  In fact, we moved it out to a clean named node under `form` to isolate it a bit, similar to how redux-form works (although I like your separation of concerns better in terms of auto-connecting and optimizing performance).

Interestingly, when we had them nested, and we followed your guide, we never provided an initial state in the formReducer (per the code above).  Now that we've moved back up to the root of state, we're getting errors that our field elements are not initialized, yet the docs are conflicted.  One doc says you don't provide an initialState in the formReducer, another does.  In fact, even another section of documentation says you don't have to provide it in modelReducer, but the examples usually do.

So clearly moving it up a level makes your system work one way, where shifting it down a tier creates some internal chaos that's more difficult to track down.

We've tried numerous combinations with setErrors, setValidity, and setFieldsErrors (in a nested state) to no avail, even with the correct model references totally explicit from the root of state.

We seem to be getting closer, but we fight with competing terminology (model means the model node, or the model leaf (field)) along with conflicting documentation, which just makes the effort of trying to figure out how to use your magic engine the right way that much harder.

We're presently hacking back from our architecture to give the library one more shot, removing our embedded nodes and putting them back at the root of state.  I wish the documented `.errors` prop was functional under `Form` like `validators`, but it's not, so we'll have to write our own and see if the systems magically starts to validate properly.

We're persistent, but sadly, if things don't start to fall into place soon, we'll be forced to go down another direction.

**Comment by davidkpiano** [comment]
Sorry to hear that. I'm working on getting the `errors` prop working with `<Form>`, and that should be landing in a patch soon.

It's generally a good practice to keep the state as flat as possible anyway, but there's no reason React Redux Form shouldn't be able to accommodate your use-case, especially since it works well with dynamic forms. I'll admit though, your use case is pretty complex, especially with dealing with hundreds of dynamic forms. If you can simplify your use case in a clear (and concise) way, I may be able to help you out further.

For the initial state, a field doesn't exist until it's acted upon, unless it's given in the `initialState` of the `formReducer`.

I'm definitely open to areas where you would like to see the docs improved. I'm working hard to make this v1.0 ready, and don't want to commit to docs until I'm 100% sure a feature in <0.x will see the light of day in v1.0.
